File: Tests/Tests/STPPaymentIntentTest.m
//
//  STPPaymentIntentTest.m
//  StripeiOS Tests
//
//  Created by Daniel Jackson on 6/27/18.
//  Copyright © 2018 Stripe, Inc. All rights reserved.
//

#import <XCTest/XCTest.h>

#import "STPPaymentIntent.h"
#import "STPPaymentIntent+Private.h"

#import "STPFixtures.h"
#import "STPTestUtils.h"

@interface STPPaymentIntentTest : XCTestCase

@end

@implementation STPPaymentIntentTest

- (void)testIdentifierFromSecret {
    XCTAssertEqualObjects([STPPaymentIntent idFromClientSecret:@"pi_123_secret_XYZ"],
                          @"pi_123");
    XCTAssertEqualObjects([STPPaymentIntent idFromClientSecret:@"pi_123_secret_RandomlyContains_secret_WhichIsFine"],
                          @"pi_123");

    XCTAssertNil([STPPaymentIntent idFromClientSecret:@""]);
    XCTAssertNil([STPPaymentIntent idFromClientSecret:@"po_123_secret_HasBadPrefix"]);
    XCTAssertNil([STPPaymentIntent idFromClientSecret:@"MissingSentinalForSplitting"]);
}

- (void)testStatusFromString {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated"
    XCTAssertEqual(STPPaymentIntentStatusRequiresSourceAction, STPPaymentIntentStatusRequiresAction);
#pragma clang diagnostic pop
    XCTAssertEqual([STPPaymentIntent statusFromString:@"requires_payment_method"],
                   STPPaymentIntentStatusRequiresPaymentMethod);
    XCTAssertEqual([STPPaymentIntent statusFromString:@"REQUIRES_PAYMENT_METHOD"],
                   STPPaymentIntentStatusRequiresPaymentMethod);

    XCTAssertEqual([STPPaymentIntent statusFromString:@"requires_confirmation"],
                   STPPaymentIntentStatusRequiresConfirmation);
    XCTAssertEqual([STPPaymentIntent statusFromString:@"REQUIRES_CONFIRMATION"],
                   STPPaymentIntentStatusRequiresConfirmation);

    XCTAssertEqual([STPPaymentIntent statusFromString:@"requires_action"],
                   STPPaymentIntentStatusRequiresAction);
    XCTAssertEqual([STPPaymentIntent statusFromString:@"REQUIRES_ACTION"],
                   STPPaymentIntentStatusRequiresAction);

    XCTAssertEqual([STPPaymentIntent statusFromString:@"processing"],
                   STPPaymentIntentStatusProcessing);
    XCTAssertEqual([STPPaymentIntent statusFromString:@"PROCESSING"],
                   STPPaymentIntentStatusProcessing);

    XCTAssertEqual([STPPaymentIntent statusFromString:@"succeeded"],
                   STPPaymentIntentStatusSucceeded);
    XCTAssertEqual([STPPaymentIntent statusFromString:@"SUCCEEDED"],
                   STPPaymentIntentStatusSucceeded);

    XCTAssertEqual([STPPaymentIntent statusFromString:@"requires_capture"],
                   STPPaymentIntentStatusRequiresCapture);
    XCTAssertEqual([STPPaymentIntent statusFromString:@"REQUIRES_CAPTURE"],
                   STPPaymentIntentStatusRequiresCapture);

    XCTAssertEqual([STPPaymentIntent statusFromString:@"canceled"],
                   STPPaymentIntentStatusCanceled);
    XCTAssertEqual([STPPaymentIntent statusFromString:@"CANCELED"],
                   STPPaymentIntentStatusCanceled);

    XCTAssertEqual([STPPaymentIntent statusFromString:@"garbage"],
                   STPPaymentIntentStatusUnknown);
    XCTAssertEqual([STPPaymentIntent statusFromString:@"GARBAGE"],
                   STPPaymentIntentStatusUnknown);
}

- (void)testCaptureMethodFromString {
    XCTAssertEqual([STPPaymentIntent captureMethodFromString:@"manual"],
                   STPPaymentIntentCaptureMethodManual);
    XCTAssertEqual([STPPaymentIntent captureMethodFromString:@"MANUAL"],
                   STPPaymentIntentCaptureMethodManual);

    XCTAssertEqual([STPPaymentIntent captureMethodFromString:@"automatic"],
                   STPPaymentIntentCaptureMethodAutomatic);
    XCTAssertEqual([STPPaymentIntent captureMethodFromString:@"AUTOMATIC"],
                   STPPaymentIntentCaptureMethodAutomatic);

    XCTAssertEqual([STPPaymentIntent captureMethodFromString:@"garbage"],
                   STPPaymentIntentCaptureMethodUnknown);
    XCTAssertEqual([STPPaymentIntent captureMethodFromString:@"GARBAGE"],
                   STPPaymentIntentCaptureMethodUnknown);
}

- (void)testConfirmationMethodFromString {
    XCTAssertEqual([STPPaymentIntent confirmationMethodFromString:@"secret"],
                   STPPaymentIntentConfirmationMethodSecret);
    XCTAssertEqual([STPPaymentIntent confirmationMethodFromString:@"SECRET"],
                   STPPaymentIntentConfirmationMethodSecret);

    XCTAssertEqual([STPPaymentIntent confirmationMethodFromString:@"publishable"],
                   STPPaymentIntentConfirmationMethodPublishable);
    XCTAssertEqual([STPPaymentIntent confirmationMethodFromString:@"PUBLISHABLE"],
                   STPPaymentIntentConfirmationMethodPublishable);

    XCTAssertEqual([STPPaymentIntent confirmationMethodFromString:@"garbage"],
                   STPPaymentIntentConfirmationMethodUnknown);
    XCTAssertEqual([STPPaymentIntent confirmationMethodFromString:@"GARBAGE"],
                   STPPaymentIntentConfirmationMethodUnknown);
}

- (void)testSetupFutureUsageFromString {
    XCTAssertEqual([STPPaymentIntent setupFutureUsageFromString:@"on_session"],
                   STPPaymentIntentSetupFutureUsageOnSession);
    XCTAssertEqual([STPPaymentIntent setupFutureUsageFromString:@"ON_SESSION"],
                   STPPaymentIntentSetupFutureUsageOnSession);

    XCTAssertEqual([STPPaymentIntent setupFutureUsageFromString:@"off_session"],
                   STPPaymentIntentSetupFutureUsageOffSession);
    XCTAssertEqual([STPPaymentIntent setupFutureUsageFromString:@"OFF_SESSION"],
                   STPPaymentIntentSetupFutureUsageOffSession);
    
    XCTAssertEqual([STPPaymentIntent setupFutureUsageFromString:@"garbage"],
                   STPPaymentIntentSetupFutureUsageUnknown);
    XCTAssertEqual([STPPaymentIntent setupFutureUsageFromString:@"GARBAGE"],
                   STPPaymentIntentSetupFutureUsageUnknown);
}

#pragma mark - Description Tests

- (void)testDescription {
    STPPaymentIntent *paymentIntent = [STPFixtures paymentIntent];

    XCTAssertNotNil(paymentIntent);
    NSString *desc = paymentIntent.description;
    XCTAssertTrue([desc containsString:NSStringFromClass([paymentIntent class])]);
    XCTAssertGreaterThan(desc.length, 500UL, @"Custom description should be long");
}

#pragma mark - STPAPIResponseDecodable Tests

- (void)testDecodedObjectFromAPIResponseRequiredFields {
    NSDictionary *fullJson = [STPTestUtils jsonNamed:STPTestJSONPaymentIntent];

    XCTAssertNotNil([STPPaymentIntent decodedObjectFromAPIResponse:fullJson], @"can decode with full json");

    NSArray<NSString *> *requiredFields = @[
                                            @"id",
                                            @"client_secret",
                                            @"amount",
                                            @"currency",
                                            @"livemode",
                                            @"status",
                                            ];

    for (NSString *field in requiredFields) {
        NSMutableDictionary *partialJson = [fullJson mutableCopy];

        XCTAssertNotNil(partialJson[field], @"json should contain %@", field);
        [partialJson removeObjectForKey:field];

        XCTAssertNil([STPPaymentIntent decodedObjectFromAPIResponse:partialJson], @"should not decode without %@", field);
    }
}

- (void)testDecodedObjectFromAPIResponseMapping {
    NSDictionary *response = [STPTestUtils jsonNamed:@"PaymentIntent"];
    STPPaymentIntent *paymentIntent = [STPPaymentIntent decodedObjectFromAPIResponse:response];

    XCTAssertEqualObjects(paymentIntent.stripeId, @"pi_1Cl15wIl4IdHmuTbCWrpJXN6");
    XCTAssertEqualObjects(paymentIntent.clientSecret, @"pi_1Cl15wIl4IdHmuTbCWrpJXN6_secret_EkKtQ7Sg75hLDFKqFG8DtWcaK");
    XCTAssertEqualObjects(paymentIntent.amount, @2345);
    XCTAssertEqualObjects(paymentIntent.canceledAt, [NSDate dateWithTimeIntervalSince1970:1530911045]);
    XCTAssertEqual(paymentIntent.captureMethod, STPPaymentIntentCaptureMethodManual);
    XCTAssertEqual(paymentIntent.confirmationMethod, STPPaymentIntentConfirmationMethodPublishable);
    XCTAssertEqualObjects(paymentIntent.created, [NSDate dateWithTimeIntervalSince1970:1530911040]);
    XCTAssertEqualObjects(paymentIntent.currency, @"usd");
    XCTAssertEqualObjects(paymentIntent.stripeDescription, @"My Sample PaymentIntent");
    XCTAssertFalse(paymentIntent.livemode);
    XCTAssertEqualObjects(paymentIntent.receiptEmail, @"danj@example.com");
    
    // Deprecated: `nextSourceAction` & `authorizeWithURL` should just be aliases for `nextAction` & `redirectToURL`
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated"
    XCTAssertEqual(paymentIntent.nextSourceAction, paymentIntent.nextAction, @"Should be the same object.");
    XCTAssertEqual(paymentIntent.nextSourceAction.authorizeWithURL, paymentIntent.nextAction.redirectToURL, @"Should be the same object.");
#pragma clang diagnostic pop

    // nextAction
    XCTAssertNotNil(paymentIntent.nextAction);
    XCTAssertEqual(paymentIntent.nextAction.type, STPIntentActionTypeRedirectToURL);
    XCTAssertNotNil(paymentIntent.nextAction.redirectToURL);
    XCTAssertNotNil(paymentIntent.nextAction.redirectToURL.url);
    NSURL *returnURL = paymentIntent.nextAction.redirectToURL.returnURL;
    XCTAssertNotNil(returnURL);
    XCTAssertEqualObjects(returnURL, [NSURL URLWithString:@"payments-example://stripe-redirect"]);
    NSURL *url = paymentIntent.nextAction.redirectToURL.url;
    XCTAssertNotNil(url);

    XCTAssertEqualObjects(url, [NSURL URLWithString:@"https://hooks.stripe.com/redirect/authenticate/src_1Cl1AeIl4IdHmuTb1L7x083A?client_secret=src_client_secret_DBNwUe9qHteqJ8qQBwNWiigk"]);
    XCTAssertEqualObjects(paymentIntent.sourceId, @"src_1Cl1AdIl4IdHmuTbseiDWq6m");
    XCTAssertEqual(paymentIntent.status, STPPaymentIntentStatusRequiresAction);
    XCTAssertEqual(paymentIntent.setupFutureUsage, STPPaymentIntentSetupFutureUsageNone);
    
    XCTAssertEqualObjects(paymentIntent.paymentMethodTypes, @[@(STPPaymentMethodTypeCard)]);
    
    // lastPaymentError
    
    XCTAssertNotNil(paymentIntent.lastPaymentError);
    XCTAssertEqualObjects(paymentIntent.lastPaymentError.code, @"payment_intent_authentication_failure");
    XCTAssertEqualObjects(paymentIntent.lastPaymentError.docURL, @"https://stripe.com/docs/error-codes/payment-intent-authentication-failure");
    XCTAssertEqualObjects(paymentIntent.lastPaymentError.message, @"The provided PaymentMethod has failed authentication. You can provide payment_method_data or a new PaymentMethod to attempt to fulfill this PaymentIntent again.");
    XCTAssertNotNil(paymentIntent.lastPaymentError.paymentMethod);
    XCTAssertEqual(paymentIntent.lastPaymentError.type, STPPaymentIntentLastPaymentErrorTypeInvalidRequest);

    XCTAssertNotEqual(paymentIntent.allResponseFields, response, @"should have own copy of fields");
    XCTAssertEqualObjects(paymentIntent.allResponseFields, response, @"fields values should match");
}

@end

File: Tests/Tests/STPPaymentIntentTest.swift
//
//  STPPaymentIntentTest.swift
//  StripeiOS Tests
//
//  Created by Daniel Jackson on 6/27/18.
//  Copyright © 2018 Stripe, Inc. All rights reserved.
//

@testable import Stripe

class STPPaymentIntentTest: XCTestCase {
  func testIdentifierFromSecret() {
    XCTAssertEqual(
      STPPaymentIntent.id(fromClientSecret: "pi_123_secret_XYZ"),
      "pi_123")
    XCTAssertEqual(
      STPPaymentIntent.id(fromClientSecret: "pi_123_secret_RandomlyContains_secret_WhichIsFine"),
      "pi_123")

    XCTAssertNil(STPPaymentIntent.id(fromClientSecret: ""))
    XCTAssertNil(STPPaymentIntent.id(fromClientSecret: "po_123_secret_HasBadPrefix"))
    XCTAssertNil(STPPaymentIntent.id(fromClientSecret: "MissingSentinalForSplitting"))
  }

  // MARK: - Description Tests
  func testDescription() {
    let paymentIntent = STPFixtures.paymentIntent()

    XCTAssertNotNil(paymentIntent)
    let desc = paymentIntent!.description
    XCTAssertTrue(desc.contains(NSStringFromClass(type(of: paymentIntent!).self)))
    XCTAssertGreaterThan((desc.count), 500, "Custom description should be long")
  }

  // MARK: - STPAPIResponseDecodable Tests
  func testDecodedObjectFromAPIResponseRequiredFields() {
    let fullJson = STPTestUtils.jsonNamed(STPTestJSONPaymentIntent)

    XCTAssertNotNil(
      STPPaymentIntent.decodedObject(fromAPIResponse: fullJson), "can decode with full json")

    let requiredFields = ["id", "client_secret", "amount", "currency", "livemode", "status"]

    for field in requiredFields {
      var partialJson = fullJson

      XCTAssertNotNil(partialJson?[field])
      partialJson?.removeValue(forKey: field)

      XCTAssertNil(STPPaymentIntent.decodedObject(fromAPIResponse: partialJson))
    }
  }

  func testDecodedObjectFromAPIResponseMapping() {
    let response = STPTestUtils.jsonNamed("PaymentIntent")!
    let paymentIntent = STPPaymentIntent.decodedObject(fromAPIResponse: response)!

    XCTAssertEqual(paymentIntent.stripeId, "pi_1Cl15wIl4IdHmuTbCWrpJXN6")
    XCTAssertEqual(
      paymentIntent.clientSecret, "pi_1Cl15wIl4IdHmuTbCWrpJXN6_secret_EkKtQ7Sg75hLDFKqFG8DtWcaK")
    XCTAssertEqual(paymentIntent.amount, 2345)
    XCTAssertEqual(paymentIntent.canceledAt, Date(timeIntervalSince1970: 1_530_911_045))
    XCTAssertEqual(paymentIntent.captureMethod, .manual)
    XCTAssertEqual(paymentIntent.confirmationMethod, .automatic)
    XCTAssertEqual(paymentIntent.created, Date(timeIntervalSince1970: 1_530_911_040))
    XCTAssertEqual(paymentIntent.currency, "usd")
    XCTAssertEqual(paymentIntent.stripeDescription, "My Sample PaymentIntent")
    XCTAssertFalse(paymentIntent.livemode)
    XCTAssertEqual(paymentIntent.receiptEmail, "danj@example.com")

    // Deprecated: `nextSourceAction` & `authorizeWithURL` should just be aliases for `nextAction` & `redirectToURL`
    //#pragma clang diagnostic push
    //#pragma clang diagnostic ignored "-Wdeprecated"
    XCTAssertEqual(
      paymentIntent.nextSourceAction, paymentIntent.nextAction, "Should be the same object.")
    XCTAssertEqual(
      paymentIntent.nextSourceAction!.authorizeWithURL!, paymentIntent.nextAction!.redirectToURL,
      "Should be the same object.")
    //#pragma clang diagnostic pop

    // nextAction
    XCTAssertNotNil(paymentIntent.nextAction)
    XCTAssertEqual(paymentIntent.nextAction!.type, .redirectToURL)
    XCTAssertNotNil(paymentIntent.nextAction!.redirectToURL)
    XCTAssertNotNil(paymentIntent.nextAction!.redirectToURL!.url)
    let returnURL = paymentIntent.nextAction!.redirectToURL!.returnURL
    XCTAssertNotNil(returnURL)
    XCTAssertEqual(returnURL, URL(string: "payments-example://stripe-redirect"))
    let url = paymentIntent.nextAction!.redirectToURL!.url
    XCTAssertNotNil(url)

    XCTAssertEqual(
      url,
      URL(
        string:
          "https://hooks.stripe.com/redirect/authenticate/src_1Cl1AeIl4IdHmuTb1L7x083A?client_secret=src_client_secret_DBNwUe9qHteqJ8qQBwNWiigk"
      ))
    XCTAssertEqual(paymentIntent.sourceId, "src_1Cl1AdIl4IdHmuTbseiDWq6m")
    XCTAssertEqual(paymentIntent.status, .requiresAction)
    XCTAssertEqual(paymentIntent.setupFutureUsage, .none)

    XCTAssertEqual(
      paymentIntent.paymentMethodTypes, [NSNumber(value: STPPaymentMethodType.card.rawValue)])

    // lastPaymentError

    XCTAssertNotNil(paymentIntent.lastPaymentError)
    XCTAssertEqual(paymentIntent.lastPaymentError!.code, "payment_intent_authentication_failure")
    XCTAssertEqual(
      paymentIntent.lastPaymentError!.docURL,
      "https://stripe.com/docs/error-codes/payment-intent-authentication-failure")
    XCTAssertEqual(
      paymentIntent.lastPaymentError!.message,
      "The provided PaymentMethod has failed authentication. You can provide payment_method_data or a new PaymentMethod to attempt to fulfill this PaymentIntent again."
    )
    XCTAssertNotNil(paymentIntent.lastPaymentError!.paymentMethod)
    XCTAssertEqual(paymentIntent.lastPaymentError!.type, .invalidRequest)

    // Shipping
    XCTAssertNotNil(paymentIntent.shipping)
    XCTAssertEqual(paymentIntent.shipping!.carrier, "USPS")
    XCTAssertEqual(paymentIntent.shipping!.name, "Dan")
    XCTAssertEqual(paymentIntent.shipping!.phone, "1-415-555-1234")
    XCTAssertEqual(paymentIntent.shipping!.trackingNumber, "xyz123abc")
    XCTAssertNotNil(paymentIntent.shipping!.address)
    XCTAssertEqual(paymentIntent.shipping!.address!.city, "San Francisco")
    XCTAssertEqual(paymentIntent.shipping!.address!.country, "USA")
    XCTAssertEqual(paymentIntent.shipping!.address!.line1, "123 Main St")
    XCTAssertEqual(paymentIntent.shipping!.address!.line2, "Apt 456")
    XCTAssertEqual(paymentIntent.shipping!.address!.postalCode, "94107")
    XCTAssertEqual(paymentIntent.shipping!.address!.state, "CA")

    XCTAssertEqual(paymentIntent.allResponseFields as NSDictionary, response as NSDictionary)
  }
}
